src/aprl/common/mujoco.py

```diff
@@ -18,9 +22,95 @@ def from_flattened(flattened, sim):
         qvel = flattened[sim.model.nq:sim.model.nq + sim.model.nv]
         return MujocoState(qpos, qvel)
 
-    def set_mjdata(self, data):
-        data.qpos[:] = self.qpos
-        data.qvel[:] = self.qvel
+    def set_mjdata(self, data, old_mujoco=False):
+        if old_mujoco:
+            data.qpos = self.qpos
+            data.qvel = self.qvel
+        else:
+            data.qpos[:] = self.qpos
+            data.qvel[:] = self.qvel
 
     def flatten(self):
-        return np.concatenate((self. qpos, self.qvel))
+        return np.concatenate((self.qpos, self.qvel))
+
+
+class ResettableEnv(gym.Env):
+    """A Gym environment that can be reset to an arbitrary state."""
+    @abstractmethod
+    def get_state(self):
+        """Returns a serialized representation of the current state."""
+        pass
+
+    @abstractmethod
+    def set_state(self, x):
+        """Restores the environment to a previously saved state.
+        :param x: return value of a previous call to get_state()."""
+        pass
+
+
+class OldMujocoResettableWrapper(ResettableEnv, MultiWrapper):
+    """Converts a MujocoEnv into a ResettableEnv.
+
+    Specifically designed to handle getting and setting states with Mujoco 1.31 / mujoco_py 0.5.7
+    Note all MuJoCo environments are resettable."""
+    def __init__(self, env):
+        """Wraps a MujocoEnv, adding get_state and set_state methods.
+        :param env: a MujocoEnv. NOTE: it must not be wrapped in a TimeLimit."""
+        if hasattr(env, '_max_episode_steps'):
+            raise TypeError('Environment must not have a time limit '
+                            '(try passing in env.unwrapped instead).')
+        gym.Wrapper.__init__(self, env)
+        self.sim = env.unwrapped.env_scene
+
+    def get_full_state(self):
+        state_dict = {}
+        for k, v in type(self.sim.data._wrapped.contents).__dict__['_fields_']:
+            if k not in ['contact', 'buffer']:
+                state_dict[k] = getattr(self.sim.data, k)
+        return state_dict
+
+    def get_state(self, all_data=False):
+        """Serializes the qpos and qvel state of the MuJoCo emulator.
+
+        :param all_data (bool) whether to return full MjData dict and env radius if it exists
+        :return: ([float], dict<str,[float]>, float) state, full_state, radius
+                 with the latter two only being returned with all_data=True
+        """
+        state = MujocoState.from_mjdata(self.sim.data).flatten()
+        if all_data:
+            full_state = self.get_full_state()
+            try:
+                radius = self.env.env.RADIUS
+            except AttributeError:
+                # this environment does not have a radius.
+                radius = None
+            return (state, full_state, radius)
```

Comment:
I did this to make the lookback state getting and setting more concise since lookback in generally verbose. But I need to handle the case of lookback being used in environments without radii more gracefully anyway so I will change this.